CONTRIBUTING.md

```diff
@@ -77,6 +77,80 @@ README for more details.
 - Once a pull request is good to go, the code reviewer is the one who merges
   it, then deletes the branch.
 
+## Controller design
+
+To maintain consistency throughout the project, we've come up with the following
+design rules for our Rails controllers:
+
+### RESTful controllers
+
+* Only use CRUD methods in controllers.
+
+* Prefer adding a new controller with one of the CRUD methods over creating a
+  custom method in an existing controller. For example, if your app allows a
+  user to update their email and their password on two different pages, instead of
+  using a single controller with methods called `update_email` and
+  `update_password`, create two controllers and name the methods `update`, i.e.
+  `EmailsController#update` and `PasswordsController#update`. See
+  http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/ for more about
+  this design pattern.
+
+### Lean controllers
+* Keep as much business logic out of controllers.
+
+* Use specialized classes to handle the operations
+  * These will be Form Objects for the most part, since
+  most of what the app does is process user input via a form submission, or
+  clicking a link in email that contains a token.
+
+* Form Object rules:
+  - Must have a single public method called `submit` that returns a [FormResponse] object.
+  - Must use ActiveModel validations to validate the user input.
+  - Must be placed in `app/forms`.
+
+* Examples of Form/Service Objects:
+  - [EmailConfirmationTokenValidator]
+  - [PasswordForm]
+
+* The basic outline of how a controller interacts with this class is:
+```ruby
+result = Class.new(user).submit(params) # this returns a FormResponse object
+# all the necessary analytics attributes are captured inside the Form Object
+analytics.track_event('Some Event Name', result.to_h)
+
+if result.success?
+  handle_success
+else
+  handle_failure
+end
+```
+
+* Only make one call to `analytics.track_event` at the beginning of the
```

Comment:
at the beginning, but after submitting the form since we don't know success/failure until after form submission.